scripts/bwctest-osd.sh

```diff
@@ -0,0 +1,180 @@
+}
+
+while getopts ":ha:p:b:v:o:d:" arg; do
+    case $arg in
+        h)
+            usage
+            exit 1
+            ;;
+        a)
+            BIND_ADDRESS=$OPTARG
+            ;;
+        p)
+            BIND_PORT=$OPTARG
+            ;;    
+        b)
+            BUNDLED_OSD=$OPTARG
+            ;;
+        v)
+            VERSIONS=$OPTARG
+            ;;
+        o)
+            OPENSEARCH=$OPTARG
+            ;;    
+        d)
+            OSD=$OPTARG
+            ;;     
+        :)
+            echo "-${OPTARG} requires an argument"
+            usage
+            exit 1
+            ;;
+        ?)
+            echo "Invalid option: -${OPTARG}"
+            exit 1
+            ;;
+    esac
+done
+
+if [ -z "$BIND_ADDRESS" ]
+then
+  BIND_ADDRESS="localhost"
+fi
+
+if [ -z "$BIND_PORT" ]
+then
+  BIND_PORT="5601"
+fi
+
+if [ -v "VERSIONS" ]
+then
+  testArray=($VERSIONS)  
+else
+  testArray=("odfe-0.10.0" "odfe-1.0.2" "odfe-1.1.0" "odfe-1.2.1" "odfe-1.3.0" "odfe-1.4.0" "odfe-1.7.0" "odfe-1.8.0" "odfe-1.9.0" "odfe-1.11.0" "odfe-1.13.2" "osd-1.0.0" "osd-1.1.0")  
+fi
+
+if [ -z "$BUNDLED_OSD" ]
+then
+  BUNDLED_OSD="false"    
+fi
+
+if [ $BUNDLED_OSD == "false" ]
+then
+  osdtype="osd"
+else
+  osdtype="osd-bundle"
+fi    
+
+#define path
+cwd=$(pwd)
+dir="$cwd/bwc-tmp"
+testdir="$dir/test"
+if [ -d "$dir" ]; then
+  echo "bwc-tmp exists and needs to be removed"   
+  rm -rf "$dir"
+fi
+mkdir "$dir"
+mkdir "$testdir"
+
+#unzip opensearch and dashboards
+cp $OPENSEARCH $dir
+cp $OSD $dir
+cd $dir
+IFS='/' read -ra ADDR <<< "$OPENSEARCH"
+opensearchtar=${ADDR[-1]}
+IFS='.' read -ra ADDR <<< "$opensearchtar"
+opensearch=${ADDR[0]}
+
+IFS='/' read -ra ADDR <<< "$OSD"
+osdtar=${ADDR[-1]}
+IFS='.' read -ra ADDR <<< "$osdtar"
+osd=${ADDR[0]}
+
+tar -xvf $opensearchtar
+tar -xvf $osdtar
+opensearch_dir="$dir/$opensearch"
+osd_dir="$dir/$osd"
+ 
+#setup cypress test env
+echo "setup cypress test env"
+git clone https://github.com/opensearch-project/opensearch-dashboards-functional-test "$testdir"
+rm -rf "$testdir/cypress/integration"
+cp -r "$cwd/cypress/integration" "$testdir/cypress"
+cd "$testdir"
+npm install
+
+#Run OpenSearch and Dashboards
+cd "$testdir"
+for i in ${!testArray[@]}; 
+do
+  version=${testArray[$i]}
+  #setup opensearch
+  #copy and unzip data in opensearch
+  echo "---------------set up opensearch env for $version---------------"
+  rm -rf "$opensearch_dir/data"
+  cd $opensearch_dir
+  cp "$cwd/cypress/test-data/$osdtype/$version.tar.gz" . 
+  tar -xvf "$opensearch_dir/$version.tar.gz" >> /dev/null 2>&1
+  rm "$version.tar.gz"
+  cd "$testdir"
+  echo "ready to test"
+  #run bwc tests
+  echo "---------------run bwc test for $version -----------------"
+  {
+    cd "$opensearch_dir"
+    ./bin/opensearch
+  } >> /dev/null 2>&1 &
+  {
+    sleep 10
+    cd "$osd_dir"
+    ./bin/opensearch-dashboards
+  } >> /dev/null 2>&1 & 
+  sleep 20
```

Comment:
Why do we need to sleep at these points?

Reply:
IIUIC, I think the original intention is to hold the thread and  wait for `opensearch-dashboards` to get startup. Same thing on `line 153` for waiting `opensearch` engine to be ready. Usually it is not a good practice to use thread `sleep` in production code for synchronization purpose, because there is no guarantee the previous job is done within `10s` or `20s`, and you may lead into an unexpected behavior.  I am not familiar with bash multi-threading but just wondering is there a way to check engine/dashboard is fully up? or it may not be a big deal for internal testing script?  Love to hear some discussion from team.